driver/rows.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020-2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package driver
+
+import (
+	"database/sql/driver"
+	"encoding/json"
+	"io"
+	"reflect"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/vitess/go/vt/proto/query"
+)
+
+// Rows is an iterator over an executed query's results.
+type Rows struct {
+	ctx  *sql.Context
+	cols sql.Schema
+	rows sql.RowIter
+}
+
+// Columns returns the names of the columns. The number of
+// columns of the result is inferred from the length of the
+// slice. If a particular column name isn't known, an empty
+// string should be returned for that entry.
+func (r *Rows) Columns() []string {
+	names := make([]string, len(r.cols))
+	for i, col := range r.cols {
+		names[i] = col.Name
+	}
+	return names
+}
+
+// Close closes the rows iterator.
+func (r *Rows) Close() error {
+	return r.rows.Close(r.ctx)
+}
+
+// Next is called to populate the next row of data into
+// the provided slice. The provided slice will be the same
+// size as the Columns() are wide.
+//
+// Next should return io.EOF when there are no more rows.
+//
+// The dest should not be written to outside of Next. Care
+// should be taken when closing Rows not to modify
+// a buffer held in dest.
+func (r *Rows) Next(dest []driver.Value) error {
+	row, err := r.rows.Next()
+	if err != nil {
+		return err
+	}
+	if len(row) == 0 {
+		return nil
+	}
+
+	if _, ok := row[0].(sql.OkResult); ok {
```

Comment:
The engine really expects you to consume all of an iterator before calling Close() on it. So the best thing to do would be to skip over any OkResults here and wait for the iter to give io.Eof itself.

Reply:
👍